### PR TITLE
file.managed: Allow local file sources to use source_hash

### LINKED ISSUE #26526: Managing a file:// source fails
To reproduce:
 1. Created a file `/tmp/a` and added some garbage content.
 2. Created the following state:
```
/tmp/b:
  file.managed:
   - source: file:///tmp/a
   - source_hash: md5=4f4748076f2f9b71dcb672604e7bca0c
```
Executing the state yields
```
          ID: /tmp/b
    Function: file.managed
      Result: False
     Comment: An exception occurred in this state: Traceback (most recent call last):
                File "/usr/lib/python2.7/dist-packages/salt/state.py", line 1560, in call
                  **cdata['kwargs'])
                File "/usr/lib/python2.7/dist-packages/salt/states/file.py", line 1423, in managed
                  **kwargs
                File "/usr/lib/python2.7/dist-packages/salt/modules/file.py", line 3147, in check_managed_changes
                  group, mode, saltenv, template, contents)
                File "/usr/lib/python2.7/dist-packages/salt/modules/file.py", line 3180, in check_file_meta
                  lstats = stats(name, hash_type=source_sum.get('hash_type', None), follow_symlinks=False)
              AttributeError: 'str' object has no attribute 'get'
     Started: 15:51:56.549278
    Duration: 4.289 ms
     Changes:
```
My version:
```bash
$ salt --versions-report
                  Salt: 2015.5.3
                Python: 2.7.3 (default, Jun 22 2015, 19:33:41)
                Jinja2: 2.6
              M2Crypto: 0.21.1
        msgpack-python: 0.1.10
          msgpack-pure: Not Installed
              pycrypto: 2.4.1
               libnacl: Not Installed
                PyYAML: 3.10
                 ioflo: Not Installed
                 PyZMQ: 13.0.0
                  RAET: Not Installed
                   ZMQ: 3.2.2
                  Mako: Not Installed
               Tornado: Not Installed
 Debian source package: 2015.5.3+ds-1precise2
```

This came out of #26497.

### PR DESCRIPTION
**Pull request by terminalmage**
Fixes #26526.

### CODE REVIEW

**Comment by terminalmage** [comment]
While the diff may not be 100% clear on this, all this PR does is move the check if the file is a local file below where source_hash is evaluated, so that using a sha256 hash is done as a fallback if source_hash is not provided.

File: salt/modules/file.py [hunk 1]
```diff
@@ -3341,13 +3341,6 @@ def get_managed(
...
                 return '', {}, 'Source file {0} not found'.format(source)
-        # if its a local file
-        elif urlparsed_source.scheme == 'file':
-            file_sum = get_hash(urlparsed_source.path, form='sha256')
-            source_sum = {'hsum': file_sum, 'hash_type': 'sha256'}
-        elif source.startswith('/'):
-            file_sum = get_hash(source, form='sha256')
-            source_sum = {'hsum': file_sum, 'hash_type': 'sha256'}
         elif source_hash:
             protos = ('salt', 'http', 'https', 'ftp', 'swift', 's3')
```
Comment (lorengordon) [inline salt/modules/file.py]: @terminalmage, can we add `'file'` to this list of accepted source_hash protos? Will that allow us to specify a hash in the form: `source_hash: file:///path/to/hashfile.hash`?

**Comment by cachedout** [comment]
Makes sense. Cool.

**Comment by lorengordon** [comment]
Maybe a slight tangent, but I think [this mesage](https://github.com/saltstack/salt/pull/29289/files#diff-7c48191b14cdf053cf954d03cc317963R3355) is not quite right when the `source_hash` scheme matches the protos list. The message seems to be saying the syntax is something like `source_hash: md5=salt://path/to/hashfile.md5`. But the code passes `source_hash` without modification directly to `_urlparse()`, so I think the message is wrong. The syntax would just be `source_hash: <proto>://path/to/hashfile`. Perhaps the message is intending to explain how `extract_hash()` reads the file, but I don't think that is coming across. I think that is covered in the [file state docs](https://docs.saltstack.com/en/latest/ref/states/all/salt.states.file.html). Also, it would maybe be nice if the message referenced the allowed `protos`.

File: salt/modules/file.py (hunk 1 above)
Comment (terminalmage) [inline salt/modules/file.py]: Yeah, that would work. If you want to submit the PR, go ahead, otherwise I can do it.

**Comment by lorengordon** [comment]
Oops, already merged. Shucks.

File: salt/modules/file.py (hunk 1 above)
Comment (lorengordon) [inline salt/modules/file.py]: Alright, I'm on it.

**Comment by cachedout** [comment]
@lorengordon Can you send over a PR with those changes? I'd love to include them.

**Comment by lorengordon** [comment]
@cachedout, yep, see #29305.

**Comment by cachedout** [comment]
@lorengordon Thank you very much.

**Comment by terminalmage** [comment]
@lorengordon Thanks!

**Pull request merged**